google/cloud/spanner/keys.h

```diff
@@ -202,8 +246,65 @@ class KeySet {
   bool IsAll() const { return all_; }
 
  private:
+  template <typename RowType>
+  friend class KeySetBuilder;
+
+  friend ::google::spanner::v1::KeySet internal::ToProto(KeySet const& keyset);
+
   explicit KeySet(bool all) : all_(all) {}
 
+  class ValueRow {
+   public:
+    template <std::size_t SIZE>
+    explicit ValueRow(std::array<Value, SIZE> const& column_values)
+        : column_values_(
+              std::vector<Value>(column_values.begin(), column_values.end())) {}
+
+    std::vector<Value> const& column_values() const { return column_values_; }
+
+   private:
+    std::vector<Value> column_values_;
+  };
```

Comment:
I wonder if this `ValueRow` class can be an alias like `using ValueRow = std::vector<Value>;`, and move the `std::array` -> `std::vector` conversion logic to the "AddRow" function, like:

```cc
using ValueRow = std::vector<Value>;
template <std::size_t N>
void AddRow(std::array<Value, N> a) {
  key_values_.emplace_back(
      std::make_move_iterator(a.begin()),
      std::make_move_iterator(a.end()));
}
```

One benefit of this would be that the `KeySetBuilder` class does not need to mention the `ValueRow` type. Instead, it simply passes its array to `AddRow` and the `KeySet` handles the conversion.

I haven't looked at the other internal classes, but maybe they could be implemented similarly?

What do you think?

Reply:
So right now, KeySet has a very small public API and all its nested classes are private. ValueRow does not show up in KeySetBuilder's API at all, only inside the implementation of Build(). I'm not sure I understand what the value proposition is in this case.

Reply:
Refactored such that KeySetBuilder (as a friend) only uses member functions of KeySet and not its internal types.

Reply:
Extended refactoring to remove KeySetBuilder as a friend and added a templated constructor to KeySet.